Segmented signing checks on extensions

carlopi opened the pull request:
Bottleneck of sha256 computation is alleviated by computing sha256 on fixed-size segments, and then combining the hashes. This is basically a 2-level Merkle tree, with root, no internal nodes, and file_size / 1MB leafs.

Note that 2-level sha256 computation is computed both on threaded and non-threaded versions, since results will have to agree (and in any case the second step has constant negligible cost).

carlopi committed:
Extension signing: Separate hash computation to external script

No actual changes, just a refactor

carlopi committed:
Extension signing: Segmented Sha256 variant

Sha256 is computed chunk by chunk (2^^20 bytes ~ 1MB in size), then they
are concatenated and Sha256 is computed once again

This is basically using a 2-level Merkle tree the root, no internal nodes
and (file_size/1MB) leaf nodes.

https://en.wikipedia.org/wiki/Merkle_tree

Mytherin commented on src/main/extension/extension_load.cpp:
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
+#ifndef DUCKDB_NO_THREADS
+		std::deque<std::thread> threads;
+		for (idx_t i = 0; i < numChunks; i++)
+			threads.push_back(std::thread(computeSHA256, handle.get(), splits[i], splits[i+1], &chunks[i]));
```
nit: can we use braces on all of these loops as per the coding guidelines?

Mytherin commented on src/main/extension/extension_load.cpp:
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
 
+		const idx_t maxLenChunks = 1024 * 1024;
+		const idx_t numChunks = (signature_offset - 1) / maxLenChunks;
+		std::vector<std::string> chunks(numChunks);
```
What happens if the extension is smaller than `maxLenChunks`?

Mytherin commented on src/main/extension/extension_load.cpp:
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
-		handle->Read((void *)file_content.data(), signature_offset, 0);
+		file_content.resize(256 * numChunks);
+		for (idx_t i = 0; i < numChunks; i++) {
+			for (idx_t j = 0; j < 256; j++)
```
Can we use a `memcpy` here instead, and move `256` to a named constexpr?

carlopi replied:
Moved to reserve length + string concatenation, that should be the same but seems clearer.

Mytherin commented on src/main/extension/extension_load.cpp:
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
+			splits[i] = maxLenChunks * i;
+
+#ifndef DUCKDB_NO_THREADS
+		std::deque<std::thread> threads;
```
Can this not be `vector`?

Mytherin commented on src/main/extension/extension_load.cpp:
```diff
@@ -70,15 +84,40 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileOpener *opener, const
+		for (idx_t i = 0; i < numChunks; i++)
+			threads.push_back(std::thread(computeSHA256, handle.get(), splits[i], splits[i+1], &chunks[i]));
+
+		for (idx_t i = 0; i < numChunks; i++)
```
`for(auto &thread : threads)`

Mytherin reviewed (commented):
Thanks for the PR! Looks good. I'm not a big fan of launching threads directly in the `TryInitialLoad` - but as extensions can be loaded before the task scheduler exists (when loaded during start-up) perhaps it is the best/only way.

Some other comments:

carlopi commented:
Thanks for the various feedback!

> I'm not a big fan of launching threads directly in the TryInitialLoad - but as extensions can be loaded before the task scheduler exists (when loaded during start-up) perhaps it is the best/only way.

On threads, was very much unsure of where to go, and left it as optional commit on purpose (so that a different solution could also be rolled in).
Manual handling looked [to me] simpler here, at the very least an explanation why it went against codebase standard and your comment that task scheduler might not be there should be added.

I will fix the comments (thanks!!), check signign actually works on my fork, and then I will ping back.

carlopi committed:
Extension signing: Segmented threaded version

This brings a speed up linear on the number of threads, both on reading
AND on computing the hash

samansmink commented:
Good idea! LGTM!

Mytherin commented:
Thanks!
